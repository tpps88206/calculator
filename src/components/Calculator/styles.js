// TODO: use OOCSS

const styles = theme => ({
  functionKeys: {
    height: '20%',
  },
  digitKeys: {
    height: '80%',
  },
  calculatorKeypad: {
    height: '80%',
  },
  calculator: {
    background: 'linear-gradient(#84BAFF, #0B0E1C)',
    boxShadow: '0px 0px 20px 0px #aaa',
    height: 'inherit', // TODO: fix height value for overflow-y
  },
  keyDot: {
    fontSize: '0.75em !important',
  },
  key0: {
    borderRadius: '40px !important',
  },
});

export default styles;
