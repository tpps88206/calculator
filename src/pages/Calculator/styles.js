const styles = theme => ({
  calculator: {
    width: 480,
    height: 780,
    position: 'relative',
  },
  '@media screen and (max-width: 768px)': {
    calculator: {
      width: 360,
      height: 585,
      position: 'relative',
    },
  },
});

export default styles;
