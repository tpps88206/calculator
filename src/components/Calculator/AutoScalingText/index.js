import React, { useEffect, useRef, useState } from 'react';

import classNames from 'classnames';

import { makeStyles } from '@material-ui/styles';

import styles from './styles';

const useStyles = makeStyles(styles);

const AutoScalingText = ({ children }) => {
  const classes = useStyles();

  const [scale, setScale] = useState(1);
  const nodeRef = useRef(null);

  // TODO: 瀏覽器大小縮放時，顯示區域需要改變大小
  useEffect(() => {
    const parentNode = nodeRef.current.parentNode; // 取得顯示區域的父類別元素
    const availableWidth = parentNode.offsetWidth; // 取得父類別元素佔用的寬度
    const actualWidth = nodeRef.current.offsetWidth; // 取得顯示數字佔用的寬度
    // 使用父類別元素佔用的寬度除以顯示數字佔用的寬度
    // 如果結果大於 1 表示該空間還夠顯示所有數字
    // 如果結果小於 1 表示數字的寬度已經超過父類別
    // 需要根據超過的比例，等比例縮小
    const actualScale = availableWidth / actualWidth;

    if (scale === actualScale) return;

    if (actualScale < 1) {
      setScale(actualScale);
    } else if (scale < 1) {
      setScale(1);
    }
  });

  return (
    <div
      className={classNames(classes.autoScalingText, 'position-absolute d-inline-block')}
      style={{ transform: `scale(${scale},${scale})` }}
      ref={nodeRef}
    >
      {children}
    </div>
  );
};

export default AutoScalingText;
