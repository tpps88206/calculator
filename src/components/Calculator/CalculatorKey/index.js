import React from 'react';
import PointTarget from 'react-point';

import classNames from 'classnames';

import { makeStyles } from '@material-ui/styles';

import styles from './styles';

const useStyles = makeStyles(styles);

const CalculatorKey = ({ onPress, variant, className, ...props }) => {
  const classes = useStyles();

  return (
    <PointTarget onPoint={onPress}>
      <button
        className={classNames(
          className,
          classes.calculatorKey,
          {
            [classes.operatorKey]: variant === 'operator-key',
            [classes.digitKey]: variant === 'digit-key',
            [classes.functionKey]: variant === 'function-key',
          },
          'd-block',
          'w-100',
          'h-100',
        )}
        {...props}
      />
    </PointTarget>
  );
};

export default CalculatorKey;
