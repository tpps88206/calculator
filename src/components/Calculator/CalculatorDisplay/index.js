import React from 'react';

import { makeStyles } from '@material-ui/styles';

import AutoScalingText from '../AutoScalingText';
import styles from './styles';

const useStyles = makeStyles(styles);

const CalculatorDisplay = ({ value, ...props }) => {
  const classes = useStyles();

  return (
    <div {...props} className={classes.calculatorDisplay}>
      <AutoScalingText>{value}</AutoScalingText>
    </div>
  );
};

export default CalculatorDisplay;
