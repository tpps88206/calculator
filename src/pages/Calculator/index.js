import React, { useEffect, useRef, useState } from 'react';
import Draggable from 'react-draggable';

import classNames from 'classnames';

import Button from '@material-ui/core/Button';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Dialog from '@material-ui/core/Dialog';
import GradientIcon from '@material-ui/icons/Gradient';
import { makeStyles } from '@material-ui/styles';

import Calculator from '../../components/Calculator';
import constant from '../../constants/constant';
import styles from './styles';

const useStyles = makeStyles(styles);

const CalculatorPage = () => {
  const classes = useStyles();

  const [open, setOpen] = useState(false); // 是否開啟計算機
  const [mobile, setMobile] = useState(false); // 是否為行動裝置
  const nodeRef = useRef(null);

  useEffect(() => {
    window.addEventListener('resize', handleRWD); // 解析度改變時，檢查裝置類型和裝置大小
    handleRWD(); // 第一次進到頁面先做一次

    return () => {
      window.removeEventListener('resize', handleRWD);
    };
  }, []);

  const handleRWD = () => {
    const isMobileDevice = constant.MOBILE_LIST.some(e => navigator.userAgent.match(e));
    if (window.innerWidth < 768 && isMobileDevice) {
      setMobile(true);
    } else {
      setMobile(false);
    }
  };

  const handleClickOpen = () => {
    if (mobile) {
      document.addEventListener('click', handleOutsideClick, true);
    }
    setOpen(true);
  };

  const handleClose = () => {
    if (mobile) {
      document.removeEventListener('click', handleOutsideClick, true);
    }
    setOpen(false);
  };

  const handleOutsideClick = e => {
    if (nodeRef.current.contains(e.target)) handleClose();
  };

  const PaperComponent = ({ className, ...props }) => {
    return (
      <Draggable>
        <div className={classNames(className, classes.calculator)} {...props} />
      </Draggable>
    );
  };

  const ContentComponent = () => {
    return (
      <React.Fragment>
        <Card className="mx-2 my-2">
          <CardContent>
            <h1 id="react-calculator">React Calculator</h1>
            <p>
              <img src="https://github.com/tpps88206/calculator/workflows/CI/CD/badge.svg" alt="CI/CD" />
              <a href="https://codecov.io/gh/tpps88206/calculator">
                <img src="https://img.shields.io/codecov/c/github/tpps88206/calculator" alt="Codecov" />
              </a>
              <img src="https://img.shields.io/github/license/tpps88206/calculator" alt="License" />
            </p>
          </CardContent>
        </Card>
        <Card className="mx-2 my-2">
          <CardContent>
            <h2 id="demo">Demo</h2>
            <p>
              <a href="https://tpps88206.github.io/calculator/">https://tpps88206.github.io/calculator/</a>
            </p>
          </CardContent>
        </Card>
        <Card className="mx-2 my-2">
          <CardContent>
            <h2 id="setup">Setup</h2>
            <pre>
              <code>
                <span className="hljs-variable">$ </span>git clone <span className="hljs-symbol">https:</span>/
                <span className="hljs-regexp">/github.com/tpps</span>88206/calculator.git
                <span className="hljs-variable">$ </span>cd calculator/
                <span className="hljs-variable">$ </span>yarn install
                <span className="hljs-variable">$ </span>yarn start
              </code>
            </pre>
          </CardContent>
        </Card>
        <Card className="mx-2 my-2">
          <CardContent>
            <h2 id="technologies">Technologies</h2>
            <ul>
              <li>
                Use <strong>React Hook</strong>
              </li>
              <li>
                Develop state management with <a href="https://redux.js.org/">Redux</a> and{' '}
                <a href="https://redux-toolkit.js.org/">Redux Toolkit</a>.
              </li>
              <li>
                Develop CSS style with <a href="https://material-ui.com/styles/basics/">@material-ui/styles</a>
              </li>
              <li>
                Develop router management with <a href="https://github.com/ReactTraining/react-router">React Router</a>.
              </li>
              <li>
                Write unit test with <strong>testing-library</strong>.
              </li>
              <li>
                Use <a href="https://github.com/features/actions">Github Actions</a> to implement CI/CD and deploy on{' '}
                <strong>Github Page</strong>.
              </li>
              <li>
                Use <a href="https://material-ui.com/zh/">Material UI</a> to generate the beautiful components with
                Material Design.
              </li>
              <li>
                Use <strong>OOCSS</strong> to manage the general CSS style, it is following{' '}
                <a href="https://getbootstrap.com/">Bootstrap</a>.
              </li>
              <li>
                Use <a href="https://github.com/ReactTraining/react-point">react-point</a> to handle touch events.
              </li>
              <li>
                Use <a href="https://github.com/STRML/react-draggable">react-draggable</a> to handle Drag and Drop.
              </li>
              <li>
                Use <a href="https://lodash.com/">Lodash</a> to control logic methods.
              </li>
              <li>
                Use <a href="https://github.com/JedWatson/classnames">Classnames</a> to manage CSS class name.
              </li>
              <li>
                Use <a href="https://typicode.github.io/husky/#/">Husky</a> to handle pre-commit event.
              </li>
              <li>
                All icons are using <a href="https://material-ui.com/components/material-icons/">Material Icon</a> or{' '}
                <strong>SVG</strong>.
              </li>
              <li>
                Define coding style with <a href="https://prettier.io/">Prettier</a> and <strong>Eslint</strong>.
              </li>
            </ul>
          </CardContent>
        </Card>
        <Card className="mx-2 my-2">
          <CardContent>
            <h2 id="features">Features</h2>
            <ul>
              <li>Implement the general functions of calculator.</li>
              <li>
                In the desktop
                <ul>
                  <li>
                    The width of calculator will fix in <code>480px</code>.
                  </li>
                  <li>
                    But when the browser width is smaller than 768px, the width of calculator will fix in{' '}
                    <code>360px</code>.
                  </li>
                  <li>We can drag and drop the calculator to everywhere.</li>
                </ul>
              </li>
              <li>
                In the mobile{' '}
                <ul>
                  <li>
                    Calculator will fix below, and the height is <code>50%</code> of the browser.
                  </li>
                  <li>We can not drag and drop but we can scroll it.</li>
                </ul>
              </li>
              <li>We can open calculator with clicking button and close it with clicking outside.</li>
            </ul>
          </CardContent>
        </Card>
      </React.Fragment>
    );
  };

  return (
    <React.Fragment>
      <div className={classNames(classes.content, 'vw-100', { [classes.mobileContent]: mobile && open })} ref={nodeRef}>
        <Button
          className="mt-2 ml-2"
          variant="contained"
          color="primary"
          size="large"
          startIcon={<GradientIcon />}
          onClick={handleClickOpen}
        >
          打開計算機
        </Button>
        {ContentComponent()}
      </div>
      {mobile ? (
        open && (
          <div className={classes.paper}>
            <div className={classNames(classes.calculator, 'position-relative')}>
              <Calculator />
            </div>
          </div>
        )
      ) : (
        <Dialog open={open} onClose={handleClose} PaperComponent={PaperComponent} hideBackdrop>
          <Calculator />
        </Dialog>
      )}
    </React.Fragment>
  );
};

export default CalculatorPage;
