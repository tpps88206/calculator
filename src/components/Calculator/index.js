import React, { useEffect, useState } from 'react';

import Icon from '@material-ui/core/Icon';
import AddIcon from '@material-ui/icons/Add';
import ClearIcon from '@material-ui/icons/Clear';
import RemoveIcon from '@material-ui/icons/Remove';
import { makeStyles } from '@material-ui/styles';

import DivisionSvg from '../../assets/division.svg';
import EqualSvg from '../../assets/equal.svg';
import PercentSvg from '../../assets/percent.svg';
import ToggleSvg from '../../assets/toggle.svg';
import CalculatorDisplay from './CalculatorDisplay';
import CalculatorKey from './CalculatorKey';
import styles from './styles';

const useStyles = makeStyles(styles);

// 定義四則運算方法
const CalculatorOperations = {
  '/': (prevValue, nextValue) => prevValue / nextValue,
  '*': (prevValue, nextValue) => prevValue * nextValue,
  '+': (prevValue, nextValue) => prevValue + nextValue,
  '-': (prevValue, nextValue) => prevValue - nextValue,
  '=': (prevValue, nextValue) => nextValue,
};

const Calculator = () => {
  const classes = useStyles();

  const [value, setValue] = useState(null); // 記錄目前計算結果的數值
  const [displayValue, setDisplayValue] = useState('0'); // 記錄顯示區域需要顯示的數字
  const [operator, setOperator] = useState(null); // 記錄當前四則運算方法
  const [waitingForOperand, setWaitingForOperand] = useState(false); // 記錄是否需要將畫面的數字消除

  // TODO: 有時候按鍵會失效
  useEffect(() => {
    document.addEventListener('keydown', handleKeyDown);
    return () => {
      document.removeEventListener('keydown', handleKeyDown);
    };
  }, []);

  // 清空所有狀態
  const clearAll = () => {
    setValue(null);
    setDisplayValue('0');
    setOperator(null);
    setWaitingForOperand(false);
  };

  // 將顯示區域重新顯示 0
  const clearDisplay = () => {
    setDisplayValue('0');
  };

  // 按下鍵盤倒退鍵時，刪除最後一個數字
  const clearLastChar = () => {
    setDisplayValue(displayValue => {
      return displayValue.substring(0, displayValue.length - 1) || '0';
    });
  };

  // 正負轉換
  const toggleSign = () => {
    setDisplayValue(displayValue => {
      return String(parseFloat(displayValue) * -1);
    });
  };

  // 按下百分比
  const inputPercent = () => {
    const currentValue = parseFloat(displayValue);

    if (currentValue === 0) {
      return;
    }

    const newValue = currentValue / 100;
    setDisplayValue(String(newValue));
  };

  // 按下小數點
  const inputDot = () => {
    // 先判斷目前是否有小數點了
    if (!/\./.test(displayValue)) {
      setDisplayValue(displayValue + '.');
      // TODO: 如果 waitingForOperand 為 true 的情況下按下小數點按鍵應該顯示 0.
      setWaitingForOperand(false);
    }
  };

  // 按下數字
  const inputDigit = digit => {
    if (waitingForOperand) {
      setDisplayValue(String(digit));
      setWaitingForOperand(false);
    } else {
      setDisplayValue(displayValue => {
        return displayValue === '0' ? String(digit) : displayValue + digit;
      });
    }
  };

  // 按下四則運算
  const performOperation = nextOperator => {
    const inputValue = parseFloat(displayValue);

    if (value == null) {
      setValue(inputValue);
    } else if (operator) {
      // 如果先前已經按過其他四則運算，則先處理
      const currentValue = value || 0;
      const newValue = CalculatorOperations[operator](currentValue, inputValue);

      setValue(newValue);
      setDisplayValue(String(newValue));
    }

    setWaitingForOperand(true); // 按下四則運算後，再輸入數字的話要清空顯示區域
    setOperator(nextOperator);
  };

  const handleKeyDown = event => {
    let { key } = event;

    if (key === 'Enter') {
      key = '=';
    }

    if (/\d/.test(key)) {
      event.preventDefault();
      inputDigit(parseInt(key, 10));
    } else if (key in CalculatorOperations) {
      event.preventDefault();
      performOperation(key);
    } else if (key === '.') {
      event.preventDefault();
      inputDot();
    } else if (key === '%') {
      event.preventDefault();
      inputPercent();
    } else if (key === 'Backspace') {
      event.preventDefault();
      clearLastChar();
    } else if (key === 'Clear') {
      event.preventDefault();

      if (displayValue !== '0') {
        clearDisplay();
      } else {
        clearAll();
      }
    }
  };

  return (
    <div className={classes.calculator}>
      <CalculatorDisplay value={displayValue} />
      <div className={classes.calculatorKeypad}>
        <div className={classes.inputKeys}>
          <div className={classes.functionKeys}>
            <CalculatorKey variant="function-key" onPress={() => (displayValue !== '0' ? clearDisplay() : clearAll())}>
              {displayValue !== '0' ? 'C' : 'AC'}
            </CalculatorKey>
            <CalculatorKey variant="function-key" onPress={() => toggleSign()}>
              <Icon>
                <img src={ToggleSvg} alt="equal" width="24" height="24" />
              </Icon>
            </CalculatorKey>
            <CalculatorKey variant="function-key" onPress={() => inputPercent()}>
              <Icon>
                <img src={PercentSvg} alt="equal" width="16" height="16" />
              </Icon>
            </CalculatorKey>
          </div>
          <div className={classes.digitKeys}>
            <CalculatorKey className={classes.key0} variant="digit-key" onPress={() => inputDigit(0)}>
              0
            </CalculatorKey>
            <CalculatorKey className={classes.keyDot} variant="digit-key" onPress={() => inputDot()}>
              ●
            </CalculatorKey>
            <CalculatorKey variant="digit-key" onPress={() => inputDigit(1)}>
              1
            </CalculatorKey>
            <CalculatorKey variant="digit-key" onPress={() => inputDigit(2)}>
              2
            </CalculatorKey>
            <CalculatorKey variant="digit-key" onPress={() => inputDigit(3)}>
              3
            </CalculatorKey>
            <CalculatorKey variant="digit-key" onPress={() => inputDigit(4)}>
              4
            </CalculatorKey>
            <CalculatorKey variant="digit-key" onPress={() => inputDigit(5)}>
              5
            </CalculatorKey>
            <CalculatorKey variant="digit-key" onPress={() => inputDigit(6)}>
              6
            </CalculatorKey>
            <CalculatorKey variant="digit-key" onPress={() => inputDigit(7)}>
              7
            </CalculatorKey>
            <CalculatorKey variant="digit-key" onPress={() => inputDigit(8)}>
              8
            </CalculatorKey>
            <CalculatorKey variant="digit-key" onPress={() => inputDigit(9)}>
              9
            </CalculatorKey>
          </div>
        </div>
        <div className={classes.operatorKeys}>
          <CalculatorKey variant="operator-key" onPress={() => performOperation('/')}>
            <Icon>
              <img src={DivisionSvg} alt="division" width="16" height="16" />
            </Icon>
          </CalculatorKey>
          <CalculatorKey className={classes.keyMultiply} variant="operator-key" onPress={() => performOperation('*')}>
            <ClearIcon />
          </CalculatorKey>
          <CalculatorKey variant="operator-key" onPress={() => performOperation('-')}>
            <RemoveIcon />
          </CalculatorKey>
          <CalculatorKey variant="operator-key" onPress={() => performOperation('+')}>
            <AddIcon />
          </CalculatorKey>
          <CalculatorKey variant="operator-key" onPress={() => performOperation('=')}>
            <Icon>
              <img src={EqualSvg} alt="equal" width="16" height="16" />
            </Icon>
          </CalculatorKey>
        </div>
      </div>
    </div>
  );
};

export default Calculator;
