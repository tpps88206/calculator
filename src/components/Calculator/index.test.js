import React from 'react';

import { fireEvent, queryByAttribute, render, screen } from '../../utils/test-utils';
import Calculator from './index';

const getById = queryByAttribute.bind(null, 'id');

it('renders Calculator without crashing', () => {
  const div = document.createElement('div');
  render(<Calculator />, div);
});

it('renders Calculator with snapshot', () => {
  const { asFragment } = render(<Calculator />);
  expect(asFragment()).toMatchSnapshot();
});

it('click single button 7', () => {
  const dom = render(<Calculator />);
  const button = screen.getByText('7');
  const element = getById(dom.container, 'AutoScalingText');

  fireEvent(
    button,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );

  expect(element.textContent).toBe('7');
});

it('click 5 add 6 equals 11', () => {
  const dom = render(<Calculator />);
  const button5 = screen.getByText('5');
  const button6 = screen.getByText('6');
  const buttonAdd = getById(dom.container, 'Add');
  const buttonEquals = getById(dom.container, 'Equals');
  const element = getById(dom.container, 'AutoScalingText');

  fireEvent(
    button5,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonAdd,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    button6,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonEquals,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );

  expect(element.textContent).toBe('11');
});

it('click 5 sub 6 equals -1', () => {
  const dom = render(<Calculator />);
  const button5 = screen.getByText('5');
  const button6 = screen.getByText('6');
  const buttonSub = getById(dom.container, 'Sub');
  const buttonEquals = getById(dom.container, 'Equals');
  const element = getById(dom.container, 'AutoScalingText');

  fireEvent(
    button5,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonSub,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    button6,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonEquals,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );

  expect(element.textContent).toBe('-1');
});

it('click 5 multiply 6 equals 30', () => {
  const dom = render(<Calculator />);
  const button5 = screen.getByText('5');
  const button6 = screen.getByText('6');
  const buttonMultiply = getById(dom.container, 'Multiply');
  const buttonEquals = getById(dom.container, 'Equals');
  const element = getById(dom.container, 'AutoScalingText');

  fireEvent(
    button5,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonMultiply,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    button6,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonEquals,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );

  expect(element.textContent).toBe('30');
});

it('click 5 divided by 10 equals 0.5', () => {
  const dom = render(<Calculator />);
  const button5 = screen.getByText('5');
  const button1 = screen.getByText('1');
  const button0 = getById(dom.container, 'Key0');
  const buttonDividedBy = getById(dom.container, 'DividedBy');
  const buttonEquals = getById(dom.container, 'Equals');
  const element = getById(dom.container, 'AutoScalingText');

  fireEvent(
    button5,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonDividedBy,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    button1,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    button0,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonEquals,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );

  expect(element.textContent).toBe('0.5');
});

it('click 5 dot 6 equals 5.6', () => {
  const dom = render(<Calculator />);
  const button5 = screen.getByText('5');
  const button6 = screen.getByText('6');
  const buttonDot = getById(dom.container, 'Dot');
  const buttonEquals = getById(dom.container, 'Equals');
  const element = getById(dom.container, 'AutoScalingText');

  fireEvent(
    button5,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonDot,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    button6,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonEquals,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );

  expect(element.textContent).toBe('5.6');
});

it('click 5 and percent equals 0.05', () => {
  const dom = render(<Calculator />);
  const button5 = screen.getByText('5');
  const buttonPercent = getById(dom.container, 'Percent');
  const buttonEquals = getById(dom.container, 'Equals');
  const element = getById(dom.container, 'AutoScalingText');

  fireEvent(
    button5,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonPercent,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonEquals,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );

  expect(element.textContent).toBe('0.05');
});

it('click 5 and toggleSign equals -5', () => {
  const dom = render(<Calculator />);
  const button5 = screen.getByText('5');
  const buttonToggleSign = getById(dom.container, 'ToggleSign');
  const buttonEquals = getById(dom.container, 'Equals');
  const element = getById(dom.container, 'AutoScalingText');

  fireEvent(
    button5,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonToggleSign,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonEquals,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );

  expect(element.textContent).toBe('-5');
});

it('click 5 and reset equals 0', () => {
  const dom = render(<Calculator />);
  const button5 = screen.getByText('5');
  const buttonReset = getById(dom.container, 'Reset');
  const buttonEquals = getById(dom.container, 'Equals');
  const element = getById(dom.container, 'AutoScalingText');

  fireEvent(
    button5,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonReset,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );
  fireEvent(
    buttonEquals,
    new MouseEvent('click', {
      bubbles: true,
    }),
  );

  expect(element.textContent).toBe('0');
});
