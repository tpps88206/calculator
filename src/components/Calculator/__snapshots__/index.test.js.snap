// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Calculator with snapshot 1`] = `
<DocumentFragment>
  <div
    class="makeStyles-calculator-16 pa-1 w-100"
    id="CalculatorMainSection"
  >
    <div
      class="makeStyles-calculatorDisplay-19"
    >
      <div
        class="makeStyles-autoScalingText-20 position-absolute d-inline-block"
        id="AutoScalingText"
        style="transform: scale(1,1);"
      >
        0
      </div>
    </div>
    <div
      class="makeStyles-calculatorKeypad-15"
    >
      <div
        class="MuiGrid-root h-100 MuiGrid-container MuiGrid-spacing-xs-2"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
        >
          <div
            class="h-100"
          >
            <div
              class="makeStyles-functionKeys-13"
            >
              <div
                class="MuiGrid-root h-100 MuiGrid-container MuiGrid-spacing-xs-2"
              >
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
                >
                  <button
                    class="makeStyles-calculatorKey-24 makeStyles-functionKey-23 d-block w-100 h-100"
                    id="Reset"
                  >
                    AC
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
                >
                  <button
                    class="makeStyles-calculatorKey-24 makeStyles-functionKey-23 d-block w-100 h-100"
                    id="ToggleSign"
                  >
                    <span
                      aria-hidden="true"
                      class="material-icons MuiIcon-root"
                    >
                      <img
                        alt="equal"
                        height="36"
                        src="toggle.svg"
                        width="36"
                      />
                    </span>
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
                >
                  <button
                    class="makeStyles-calculatorKey-24 makeStyles-functionKey-23 d-block w-100 h-100"
                    id="Percent"
                  >
                    <span
                      aria-hidden="true"
                      class="material-icons MuiIcon-root"
                    >
                      <img
                        alt="equal"
                        height="32"
                        src="percent.svg"
                        width="32"
                      />
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="makeStyles-digitKeys-14 mt-1"
            >
              <div
                class="MuiGrid-root h-100 MuiGrid-container MuiGrid-spacing-xs-2"
              >
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
                >
                  <button
                    class="makeStyles-calculatorKey-24 makeStyles-digitKey-22 d-block w-100 h-100"
                  >
                    7
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
                >
                  <button
                    class="makeStyles-calculatorKey-24 makeStyles-digitKey-22 d-block w-100 h-100"
                  >
                    8
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
                >
                  <button
                    class="makeStyles-calculatorKey-24 makeStyles-digitKey-22 d-block w-100 h-100"
                  >
                    9
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
                >
                  <button
                    class="makeStyles-calculatorKey-24 makeStyles-digitKey-22 d-block w-100 h-100"
                  >
                    4
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
                >
                  <button
                    class="makeStyles-calculatorKey-24 makeStyles-digitKey-22 d-block w-100 h-100"
                  >
                    5
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
                >
                  <button
                    class="makeStyles-calculatorKey-24 makeStyles-digitKey-22 d-block w-100 h-100"
                  >
                    6
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
                >
                  <button
                    class="makeStyles-calculatorKey-24 makeStyles-digitKey-22 d-block w-100 h-100"
                  >
                    1
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
                >
                  <button
                    class="makeStyles-calculatorKey-24 makeStyles-digitKey-22 d-block w-100 h-100"
                  >
                    2
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
                >
                  <button
                    class="makeStyles-calculatorKey-24 makeStyles-digitKey-22 d-block w-100 h-100"
                  >
                    3
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8"
                >
                  <button
                    class="makeStyles-key0-18 text-left makeStyles-calculatorKey-24 makeStyles-digitKey-22 d-block w-100 h-100"
                    id="Key0"
                  >
                    0
                  </button>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
                >
                  <button
                    class="makeStyles-keyDot-17 makeStyles-calculatorKey-24 makeStyles-digitKey-22 d-block w-100 h-100"
                    id="Dot"
                  >
                    ‚óè
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <div
            class="h-100"
          >
            <div
              class="MuiGrid-root h-100 MuiGrid-container MuiGrid-spacing-xs-2"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
              >
                <button
                  class="makeStyles-calculatorKey-24 makeStyles-operatorKey-21 d-block w-100 h-100"
                  id="DividedBy"
                >
                  <span
                    aria-hidden="true"
                    class="material-icons MuiIcon-root"
                  >
                    <img
                      alt="division"
                      height="30"
                      src="division.svg"
                      width="30"
                    />
                  </span>
                </button>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
              >
                <button
                  class="makeStyles-calculatorKey-24 makeStyles-operatorKey-21 d-block w-100 h-100"
                  id="Multiply"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    style="font-size: 1em;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
              >
                <button
                  class="makeStyles-calculatorKey-24 makeStyles-operatorKey-21 d-block w-100 h-100"
                  id="Sub"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    style="font-size: 1em;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 13H5v-2h14v2z"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
              >
                <button
                  class="makeStyles-calculatorKey-24 makeStyles-operatorKey-21 d-block w-100 h-100"
                  id="Add"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root"
                    focusable="false"
                    style="font-size: 1em;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                    />
                  </svg>
                </button>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
              >
                <button
                  class="makeStyles-calculatorKey-24 makeStyles-operatorKey-21 d-block w-100 h-100"
                  id="Equals"
                >
                  <span
                    aria-hidden="true"
                    class="material-icons MuiIcon-root"
                  >
                    <img
                      alt="equal"
                      height="26"
                      src="equal.svg"
                      width="26"
                    />
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
